python/circuitsvis/attribution.py

```diff
@@ -0,0 +1,181 @@
+from jaxtyping import Float
+from typing import Optional, Union, List, Literal, Tuple
+import torch as t
+from torch import Tensor
+from transformer_lens import HookedTransformer, ActivationCache
+import einops
+from circuitsvis.attention import attention_heads, attention_patterns
+
+
+def concat_lists(*lists):
+    return [x for l in lists for x in l]
+
+
+def diagonalise(tensor: Tensor, dim: int):
+    '''
+    Returns a tensor with dimension `dim` duplicated, diagonal elements along these two dims
+    are given by `tensor`, and off-diags are zero.
+    '''
+    shape = tensor.shape
+    new_shape = shape[:dim] + (shape[dim],) + shape[dim:]
+    new_tensor = t.zeros(new_shape, dtype=tensor.dtype, device=tensor.device)
+    indices = concat_lists(
+        [slice(None)] * dim,
+        [range(shape[dim]), range(shape[dim])],
+        [slice(None)] * (len(shape) - dim - 1)
+    )
+    new_tensor[indices] = tensor
+    return new_tensor
+
+
+def from_cache(
+    cache: ActivationCache,
+    model: HookedTransformer,
+    tokens: Union[List[str], Float[Tensor, "seq"]],
+    resid_directions: Optional[Union[Float[Tensor, "*seq d_model"], Float[Tensor, "*seq d_vocab"]]] = None,
+    seq_pos: Optional[Union[int, List[int]]] = None,
+    mode: Literal["large", "small"] = "small",
+    include_b_U_attribution: bool = False, # This always points towards more common tokens I guess, so it does a lot of the heavy lifting for us (when we aren't using a baseline). Interesting!
+):

```

Comment:
```suggestion
) -> Union[RenderedHTML, Tuple[RenderedHTML, RenderedHTML]]:
```

Since the method's quite long